Add support for placing the FAB in different positions

DaveShuckerow opened the pull request:
- [x] Implement support for positioning the FAB inside the Scaffold
- [x] Tests
- [x] Demo in the gallery

I'd like to talk with you about the approach to be sure we're agreed that it's the right way to go before I continue with testing and the demo.

Commit:
Add support to move the fab between positions.

Commit:
Motion demo for the FAB works between center and end floating.

Commit:
Add a Material curve to the offset animation.

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Commit:
Move the fab position into an object

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Commit:
Updates to docs

Commit:
Updates to docs

Commit:
Merge branch 'fab-motion' of github.com:DaveShuckerow/flutter into fab-motion

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,19 +37,94 @@ enum _ScaffoldSlot {
 
+/// A position that the [FloatingActionButton] can dock to in a [Scaffold].
+/// 
+/// [getOffset] 
```
(not sure if you want detailed review comments yet, but if you do: looks like these docs are incomplete)

DaveShuckerow replied:
Documentation is ready.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,19 +37,94 @@ enum _ScaffoldSlot {
+  /// Centered FAB, floating at the bottom of the screen.
+  static const FabPosition centerFloat = const _CenterFloatFab();
+
+  const FabPosition(this.getOffset);
```
nit: in flutter, we always put the constructors first, that way we can tell at a glance if there's an implicit constructor. otherwise you'd have to scan the entire class to determine if it had an implicit constructor.

also, all public members should have documentation; see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-useless-documentation for some writing prompts for writing docs when it's not clear what to write.

DaveShuckerow replied:
Done.  Does this apply to constructors when there's only one unnamed constructor?

Hixie replied:
Yup.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,19 +37,94 @@ enum _ScaffoldSlot {
+  /// [bottomSheetSize] is the size of the [BottomSheet] (if available).
+  /// [textDirection] is the direction used by the Scaffold (either RTL or LTR).
+  /// [contentBottom] is the height from the top where the Scaffold's body ends.
+  /// [horizontalPadding] is padding the Scaffold would like the FAB to observe.
```
these will all end up in one paragraph; if that's not what you intended, consider prefixing them with ` * ` or putting a blank line between each one.

Also since these are just arguments, consider using back ticks instead of square brackets to refer to them.

DaveShuckerow replied:
Done.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,19 +37,94 @@ enum _ScaffoldSlot {
+  /// [textDirection] is the direction used by the Scaffold (either RTL or LTR).
+  /// [contentBottom] is the height from the top where the Scaffold's body ends.
+  /// [horizontalPadding] is padding the Scaffold would like the FAB to observe.
+  final Offset Function(Size scaffoldSize, Size fabSize, Size snackBarSize, Size bottomSheetSize, TextDirection textDirection, double contentBottom, double horizontalPadding) getOffset;
```
Let's use a typedef here so that people can indirect this without having to duplicate the entire signature

DaveShuckerow replied:
Based on discussion offline, changed this to using inheritance and taking a ScaffoldPrelayoutGeometry object as a parameter.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,19 +37,94 @@ enum _ScaffoldSlot {
+/// A position that the [FloatingActionButton] can dock to in a [Scaffold].
+/// 
+/// [getOffset] 
+class FabPosition {
```
presumably this should be abstract?

Hixie replied:
Oh, I see, you intend this not to be extended but to provide a callback.

Interesting. Can you elaborate on this? Why not just make this a class that gets extended? You do extend it as well, which is odd.

DaveShuckerow replied:
I've decided to take the inheritance approach to this class, the field is now an abstract method.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -771,6 +844,36 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+        _fabPosition = newPosition;
+        print(
+          'Moving fab from $_previousFabPosition to $_fabPosition'
+        );
```
i presume you didn't intend to leave the print

DaveShuckerow replied:
No, it was just for debugging.  I've removed all the prints.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -124,22 +200,12 @@ class _ScaffoldLayout extends MultiChildLayoutDelegate {
+      // To account for the FAB position being changed, we'll lerp between its old and new position.
+      final Offset currentFabOffset = currentFabPosition.getOffset(size, fabSize, snackBarSize, bottomSheetSize, textDirection, contentBottom, endPadding);
+      final Offset previousFabOffset = previousFabPosition.getOffset(size, fabSize, snackBarSize, bottomSheetSize, textDirection, contentBottom, endPadding);
+      final Offset fabOffset = new MaterialPointArcTween(begin: previousFabOffset, end: currentFabOffset).lerp(fabMoveProgress);
```
creating a tween each time seems unfortunate, we could just pass in the actual offset here if we moved the animation logic out to the caller

DaveShuckerow replied:
I've refactored as much of the tween creation as possible.

I still need to create the exact offset tween in the scaffold layout delegate because the positions still need to respond to changes in, for example, the bottom sheet or snackbar during the animation.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -771,6 +844,36 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+  void _moveFab(final FabPosition newPosition) {
+    void updatePosition() {
+      setState(() {
+        _previousFabPosition = _fabPosition;
```
what if the previous animation didn't complete?
we should probably set _previousFabPosition to its actual current position.

DaveShuckerow replied:
I've reworked this section to schedule animations more intelligently.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -771,6 +844,36 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+      });
+    } else {
+      updatePosition();
+    }
```
i don't understand this

DaveShuckerow replied:
I've reworked this section to schedule animations more intelligently.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1039,15 +1163,20 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+                textDirection: textDirection,
+                previousFabPosition: _previousFabPosition,
+                currentFabPosition: _fabPosition,
+                fabMoveProgress: new CurveTween(curve: Curves.ease).evaluate(_fabMoveController),
```
rather than create this tween every frame and the MaterialArcTween every frame as well, we should probably create the chain of Animations when the target position changes, then just pass in the current value of the animation each frame.

DaveShuckerow replied:
I've refactored as much of the tween creation as possible.

I still need to create the exact offset tween in the scaffold layout delegate because the positions still need to respond to changes in, for example, the bottom sheet or snackbar during the animation.

Hixie replied:
Tweens are mutable, if that helps.

Hixie commented:
The approach seems pretty solid to me.
@HansMuller and @amirh may have more informed opinions.

DaveShuckerow commented:
Thanks for the comments, @Hixie 

Hans and I have a detailed discussion planned for tomorrow afternoon.

Commit:
Fix a lint on the bottom sheet type

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,19 +37,94 @@ enum _ScaffoldSlot {
+/// A position that the [FloatingActionButton] can dock to in a [Scaffold].
+/// 
+/// [getOffset] 
+class FabPosition {
```
I'd prefer it if getOffset() was just a method, ScaffoldState provided the current geometry of its parts.

```
abstract class FabPosition {
  const FabPosition(); // to enable const constructors in subcalsses

  Offset getOffset(ScaffoldState scaffold);

  static const FabPosition centerFloat = const _CenterFloat();

  //...
}

class _CenterFloat extends FabPosition {
  const _CenterFloat();
  @override
  Offset getOffset(ScaffoldState scaffold) {
    final TextDirection textDirection = Directionality.of(scaffold.context).textDirection;
    final scaffoldSize = scaffold.size;
    final fabSize = scaffold.fabSize;
    final snackBarSize = scaffold.snackBarSize;
    final bottomSheetSize = scaffold.bottomSheetSize;
    // Or maybe just provide a scaffold.contentPadding getter whose insets
    // account for the snackbar, bottomsheet, appbar, etc.

    // Offset computing code is as before...
  }
}
```

HansMuller reviewed (commented):
We need to discuss how Scaffold parts will track the FAB's position

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -36,19 +37,94 @@ enum _ScaffoldSlot {
+  /// [textDirection] is the direction used by the Scaffold (either RTL or LTR).
+  /// [contentBottom] is the height from the top where the Scaffold's body ends.
+  /// [horizontalPadding] is padding the Scaffold would like the FAB to observe.
+  final Offset Function(Size scaffoldSize, Size fabSize, Size snackBarSize, Size bottomSheetSize, TextDirection textDirection, double contentBottom, double horizontalPadding) getOffset;
```
I'm worried that this long list of parameters might grow and get confusing over time as the FAB positioning spec changes. In addition an app developer that wants custom FAB positioning might need to depend on a completely different set of parameters.

Would it be possible to abstract away the set of inputs that are used to determine the position?
e.g have this method take no parameters, and the different implementations would keep members that allow them to get the inputs they need.

I would imagine a ScaffoldFabPosition class that keeps a reference to the ScaffoldState, and queries it for the inputs it needs (or alternatively maybe have ScaffoldState implement FabPosition?)

Construction is going to be tricky though as you end up with a dependency loop.
Maybe we can make the return value a type that can either be an Offset, or say "defer to scaffold startegy".
Then the ScaffoldState could internally construct a private FabPosition, provide it with all the dependencies, and defer to it when the "main" FabPosition says "defer to scaffold".

amirh replied:
Another thought - I can imagine use cases where we animate the return values of this function.
e.g we want the FAB positioned relative to some other UI element that is currently being animated.
We might want the return type to be able to say whether the position is being animated or not (so we will not trigger another animation if it is already animation)

HansMuller reviewed (commented):
We need to discuss how Scaffold parts will track the FAB's position

Commit:
Merge branch 'fab-motion' of github.com:DaveShuckerow/flutter into fab-motion

Commit:
Add a ScaffoldGeometry class

Commit:
Improve the documentation

Commit:
Improve the documentation

Commit:
Add a fab motion animator

Commit:
Add position and scale animations

Commit:
FAB entrance and motion animations work

Hixie commented:
@DaveShuckerow Should we close this PR while you work on this, to get it off the review queue?

Commit:
Get started on FAB motion

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Commit:
Make fab animation work properly.

Commit:
Change the fab animator to be stored in the state of the scaffold.

Commit:
Add a layout test

DaveShuckerow commented:
I haven't addressed all the comments yet, but I had a question that I'd like to ask.

I'm getting a test failure with the test `Floating action button directionality`.  The offset of the fab doesn't update in the test.  

I suspect that the `_ScaffoldLayout` layout delegate is not detecting the change in directionality, but the delegate is comparing the build context between builds.

Is BuildContext supposed to support equality comparison?

Commit:
Fix spacing being off

Hixie commented:
Comparing the BuildContext won't tell you anything useful.

You should pass the directionality to the scaffold layout. That way you can compare the directionality rather than the build context.

Commit:
Fix the entrance/exit animation test.

Commit:
Add a textDirection to the layout delegate.

DaveShuckerow commented:
That makes sense, but I suspect that approach might not scale.

With this change, the scaffold layout is taking in a text direction without ever using it except to re-layout when the direction changes.  It's only used by one of the FabPositioners to place the FAB at the end of the view.

I propose that we allow FabPositioners to diff the BuildContext and flag whether or not the change requires a re-layout.

Hixie commented:
I don't understand what doesn't scale. Can you elaborate?

> I propose that we allow FabPositioners to diff the BuildContext and flag whether or not the change requires a re-layout.

I don't understand what this means.

Commit:
Fix const constructor lint checks

DaveShuckerow commented:
> I don't understand what doesn't scale. Can you elaborate?

Being able to customize the fab's position with the build context as a parameter means that a fab positioner could pull any relevant fields off of the build context.

In this PR, the scaffold layout delegate is listening to the textDirection for changes and re-laying out, but it never actually uses the textDirection.  Only the fab positioner that sets up the end-floating position actually uses the textDirection.

I imagine something similar could happen to other features of the build context, such as the theme or the locale.  I can come discuss in-person as well.

Hixie commented:
If we want to allow the build context to be used, then we should use the builder pattern rather than the delegate pattern. The delegate pattern should not depend on anything outside of its constructor arguments, since otherwise it can't report when to update. The inheritance mechanism that works by getting inherited values through build contexts causes rebuilds.

Hixie commented:
(Always happy to chat in person.)

DaveShuckerow commented:
The only thing that's used anywhere is the textDirection, so I think this work is manageable without the build context.  Instead, we'll just provide the text direction directly.

I talked to @amirh and @HansMuller and we agree that this won't affect #14580 that Amir is working on.  I'll respond to all the rest of the comments on this change today and be ready for the next round of review.

Commit:
Add toStrings for the fab positioner/animator

Commit:
Add a toString for CurveTween

Commit:
Merge branch 'tostring' into fab-motion

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Rename the ScaffoldGeometry for FAB layout to ScaffoldPrelayoutGeometry.

Commit:
Change the fab motion demo icon to a simple add icon.

Commit:
Add tests and a custom fab positioner to the demo.

Commit:
Do not start the fab's motion animation when the fab is null.

Commit:
Adjust the code to pass the new tests.

Commit:
Rename for in response to Hans' comment.

Commit:
Revert the tabs fab demo

Commit:
Use timeDilation, and clean up the animation code a little.

Commit:
Clean up the prelayout geometry docs and ctr order

Commit:
Cleanup fab transition widget code

Commit:
Clean up comments on Scaffold, add cross-references between the two geometries

Commit:
Explain the fab motion animation scheduling better

Commit:
Add a const to the fab motion demo

Commit:
Make the fab animation never jank by keeping track of where to move the fab to in the future.

Commit:
Add a default fab positioner constant

DaveShuckerow commented:
Okay, I've addressed all the review comments.  Please take another look.

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+/// Material Design apps, whose implementations are available as static
+/// members of this class.
+abstract class FabPositioner {
+  const FabPositioner();
```
please use the following boilerplate for the docs of const abstract constructors:
```
  /// Abstract const constructor. This constructor enables subclasses to provide
  /// const constructors so that they can be used in const expressions.
```

Hixie replied:
(same comment in other places)

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+  String toString() => '$runtimeType()';
+}
+
+/// Provider of [FloatingActionButton] animations.
```
This needs much more detailed docs. See https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-useless-documentation for some writing prompts.

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
 }
 
-/// Geometry information for scaffold components.
+/// Responsible for placing the [FloatingActionButton] in the [Scaffold].
```
Maybe "Interface for objects that place..."?

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+
+  @override
+  T get value => math.max(first.value, next.value);
+}
```
seems like this would be a potentially useful general animation; should we make it public and put it in the animation library?

DaveShuckerow replied:
I think so, but I don't want it to block this PR.  I'll get a separate PR sent to expose MaxAnimation publicly.

DaveShuckerow replied:
Sent out in #14814

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+  /// 
+  /// The Scaffold will determine the fabSize if 
+  /// `Scaffold.floatingActionButton` is not null.
+  final Size fabSize;
```
we should expand "fab" here

Hixie replied:
otherwise people will think we are referring to the fabulous size.

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+  /// The [Size] of the [Scaffold]'s [FloatingActionButton] (if available).
+  /// 
+  /// The Scaffold will determine the fabSize if 
+  /// `Scaffold.floatingActionButton` is not null.
```
[Scaffold.floatingActionButton].

DaveShuckerow replied:
Done.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+
+  /// The [Size] of the [Scaffold]'s [BottomSheet] (if available).
+  /// 
+  /// The Scaffold will determine the bottomSheetSize if the Scaffold
```
[Scaffold], [bottomSheetSize]

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+  /// `Scaffold.floatingActionButton` is not null.
+  final Size fabSize;
+
+  /// The [Size] of the [Scaffold]'s [BottomSheet] (if available).
```
"if available" leads to the obvious question... what if it isn't? will this be null? zero? arbitrary?

Hixie replied:
(same comment in other places)

DaveShuckerow replied:
Done.  The field will be Size.zero in these cases.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+  /// is currently showing a [SnackBar].
+  final Size snackBarSize;
+
+  /// The [Scaffold]'s [TextDirection].
```
Scaffold doesn't have a TextDirection, so this probably needs more detail to describe what it really is...

DaveShuckerow replied:
I've clarified that it's the TextDirection of the Scaffold's BuildContext.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+}
+
+class _EndFloatFab extends FabPositioner {
+
```
trivial nit: extraneous blank line

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+  }
+}
+// Examples can assume:
+// ScaffoldGeometry scaffoldGeometry;
```
we normally put these pragmas at the top of the file after the imports.

DaveShuckerow replied:
I've pulled this comment out.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
 /// To get a [ValueNotifier] for the scaffold geometry call
 /// [Scaffold.geometryOf].
+/// 
+/// For information about the Scaffold's geometry that is used as part of
```
[Scaffold]

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -121,17 +322,27 @@ class _ScaffoldLayout extends MultiChildLayoutDelegate {
-  });
+    // for floating action button
+    @required this.horizontalPadding, 
+    @required this.previousFabPosition,
```
(throughout)

DaveShuckerow replied:
I've done this for all the public-facing names.  I've kept it in some of the private references for consistency with the original implementations.

Let me know if you'd like to not use `fab` to refer to floating action buttons at all.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -215,23 +426,23 @@ class _ScaffoldLayout extends MultiChildLayoutDelegate {
-      positionChild(_ScaffoldSlot.floatingActionButton, new Offset(fabX, fabY));
-      floatingActionButtonRect = new Offset(fabX, fabY) & fabSize;
+      
+      // To account for the FAB position being changed, we'll lerp between its old and new position.
```
let's say something like "we'll animate between" rather than lerp between, because we don't necessarily "lerp" in the sense of calling Offset.lerp.

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -259,19 +470,36 @@ class _ScaffoldLayout extends MultiChildLayoutDelegate {
-        || oldDelegate.textDirection != textDirection;
+        || oldDelegate.horizontalPadding != horizontalPadding
+        || oldDelegate.textDirection != textDirection
+        || !(fabMoveAnimation.isCompleted || fabMoveAnimation.isDismissed)
```
this is sketchy

DaveShuckerow replied:
Could you elaborate on the code smell you're seeing here?

The goal of this line is to have the layout continue when the animation is running.  Is there a better way to express this, or do you have a concern about this combination of a layout delegate and an animation?

Hixie replied:
This method only gets called when a new object is created.
By definition, a delegate should always give the same result given the same constructor arguments.
What you're doing here "cheats" this; you recreate an identical object, but because it happens to be new, we call this method and then you get to say "oh by the way, even though we're identical, let's recompute, because we might have changed our mind".

The usual solutions for this would be to either:
 - not pass any mutable state into the delegate (e.g. pass in `Offset`s, not `Animation<Offest>`s), or
- make the delegate contract have a Listenable API, the way we did with the CustomPainter delegate.

DaveShuckerow replied:
Done.  I've switched the animator to handling the offset directly rather than creating an animation to pass around.  This makes the getOffset code cleaner too, which is a nice plus.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -937,6 +1183,51 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
     return _currentBottomSheet;
   }
 
+  // FAB API
```
expand all these "fab"s

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -965,8 +1256,28 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+    // Update the fab animator, and then schedule the fab for repositioning.
+    if (widget.fabMotionAnimator != oldWidget.fabMotionAnimator) {
+      setState(() {
+        _fabMotionAnimator = widget.fabMotionAnimator ?? FabMotionAnimator.scaling;
```
no need to call setState from inside didUpdateWidget

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -978,6 +1289,9 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
     if (_currentBottomSheet != null)
       _currentBottomSheet._widget.animationController.dispose();
+    _fabMoveController.removeStatusListener(_onFabMoveStatus);
+    _fabMoveController.dispose();
```
no need to remove the listener if you immediately dispose

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -978,6 +1289,9 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
       _currentBottomSheet._widget.animationController.dispose();
+    _fabMoveController.removeStatusListener(_onFabMoveStatus);
+    _fabMoveController.dispose();
+    _fabMoveController = null;
```
no need to mark it as null, the class is defunct after this anyway

DaveShuckerow replied:
Done.

Currently, the SnackBarController is also getting set to null.  It's also an animation controller, so I have gone ahead and removed the line that was nulling it out to be consistent here.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -937,6 +1183,51 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+
+  void _moveFab(final FabPositioner newPositioner) {
+    setState(() {
+      _nextFabPositioner = newPositioner;
```
this doesn't seem to handle moving the fab while it's moving.
Suppose someone hits a button that moves the fab from left to right, then 80% of the way through the animation they hit "back". we'd want it to start from that 80% point, not from the far right.

DaveShuckerow replied:
Getting that will take a bit of extra plumbing:

 - We have to create an animation-dependent positioner tween that can find the correct offset between the old and new positioners.  This requires wiring the positioning logic through the layout phase and supporting an arbitrary level of nesting depth.

 - For scale and rotation tweens, the same concern applies, and we have to decide on a 'bridge' animation to connect the old to the new without any jump cuts.  For most cases, I imagine a Train animation would work.

I am concerned about the offset animation because we need to compute it for each fabPositioner we're moving between at each layout iteration.  If the number of positioners grows beyond the single digits, that could impact performance.  After I finish the other changes, I will explore the instant-response animation to make sure that it works correctly.

Hixie replied:
I haven't studied exactly what it would take to do this, but I do think we definitely need to do this. It hasn't historically been hard for us to do this, e.g. the shape border lerps support this fine (you can lerp from a circle to a rect, then in the middle of that lerp from the half-circle-half-rectangle shape to a circle).

Commit:
Add space after comma in the demo

Commit:
Add boilerplate dartdoc to all const constructors

Commit:
Comment improvement

Commit:
Rename 'fabSize' to 'floatingActionButtonSize'

Commit:
Rename 'fabSize' to 'floatingActionButtonSize'

Commit:
Rename 'fabSize' to 'floatingActionButtonSize'

Commit:
Clean up the prelayout geometry object's dartdoc

Commit:
Clean up the prelayout geometry object's dartdoc

Commit:
Remove extraneous comment

Commit:
Change possessive uses of Scaffold's to use dartdoc-compatible [Scaffold]'s

Commit:
Rename the horizontalFabPadding to an expansion

Commit:
Clean up controller cleanup and setState usage

Commit:
Animate instead of lerp

DaveShuckerow commented on examples/flutter_gallery/lib/demo/material/fab_motion_demo.dart:
```diff
@@ -0,0 +1,123 @@
+
+const String _explanatoryText =
+  "When the Scaffold's floating action button positioner changes,"
+  'the floating action button animates to its new position';
```
Done

DaveShuckerow commented:
I've done the small changes.  I'll make a second pass for the two large changes and the documentation expansion as soon as I can.

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Commit:
Make the fab position animation use offsets instead of animations

Commit:
Merge branch 'fab-motion' of github.com:DaveShuckerow/flutter into fab-motion

Commit:
Streamline the fab motion demo

Commit:
Set up the animator to start from a reasonable place when interrupting animations.

Commit:
Doc cleanup on the new animation interruption

Commit:
Expand some uses of fab and clean up constants

Commit:
Expand remaining public uses of fab to floating action button

Commit:
Expand remaining public uses of fab to floating action button

Commit:
Expand on the documentation for the fab positioner and animator

DaveShuckerow commented:
I've responded @Hixie's comments.

@HansMuller can you take a pass on this PR and see if everything looks sensible to you?  I'd like to know if there's anything that gives you pause or makes you ask a question.

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -443,6 +762,16 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
     final List<Widget> children = <Widget>[];
+    if (_currentAnimation.status == AnimationStatus.completed && widget.child != null) { 
+      return new ScaleTransition(
+        scale: widget.fabMotionAnimator.getScaleAnimation(parent: widget.fabMoveAnimation),
```
Looks like this scaling animation doesn't update `floatingActionButtonScale` in ScaffoldGeometry.
This should do something similar to what we do below in updateGeometryScale.

DaveShuckerow replied:
Done.

amirh commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -443,6 +762,16 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
   @override
   Widget build(BuildContext context) {
     final List<Widget> children = <Widget>[];
+    if (_currentAnimation.status == AnimationStatus.completed && widget.child != null) { 
```
What happens when the FAB is changed while it is animating to a new position? (or animating to a new position while the scale animation triggered by a FAB change is running?)

Looks like we might jump from between the different cases in this build method during an animation?

DaveShuckerow replied:
Previously, you were right about the jump.  I'm combining the animations now, which avoids the issue of a jump between them.

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Commit:
Refactor animations to broadcast the position properly.

Commit:
Add the ability to turn on and off the fab to the motion demo.

Commit:
Remove unused code

Commit:
Change the fab animator to animate even when the fab is exitting.

Commit:
Remove extra positioner.

DaveShuckerow commented:
Ready for review again.  PTAL.

Hixie commented on examples/flutter_gallery/lib/demo/material/fab_motion_demo.dart:
```diff
@@ -0,0 +1,105 @@
+  'the floating action button animates to its new position';
+
+class FabMotionDemo extends StatefulWidget {
+  static String routeName = '/material/fab-motion';
```
this should be const (or final), since i presume it's not supposed to change

DaveShuckerow replied:
Done

Hixie commented on examples/flutter_gallery/lib/demo/material/fab_motion_demo.dart:
```diff
@@ -0,0 +1,105 @@
+      }) 
+      : null;
+    final Widget scaffold = new Scaffold(
+      appBar: new AppBar(title: const Text('FAB Positioner'), bottom: const PreferredSize(preferredSize: const Size.fromHeight(48.0), child: const SizedBox())),
```
wrap the app bar

DaveShuckerow replied:
Done

Hixie commented on examples/flutter_gallery/lib/demo/material/fab_motion_demo.dart:
```diff
@@ -0,0 +1,105 @@
+        fabX = 16.0 + scaffoldGeometry.horizontalFloatingActionButtonPadding;
+        break;
+      case TextDirection.rtl:
+        fabX = scaffoldGeometry.scaffoldSize.width - scaffoldGeometry.floatingActionButtonSize.width - 16.0 - scaffoldGeometry.horizontalFloatingActionButtonPadding;
```
we prefer listing rtl first, see https://master-docs-flutter-io.firebaseapp.com/flutter/intl/TextDirection-class.html

DaveShuckerow replied:
Done

Hixie commented on examples/flutter_gallery/lib/gallery/item.dart:
```diff
@@ -158,6 +158,13 @@ List<GalleryItem> _buildGalleryItems() {
+      category: 'Material Components',
+      routeName: FabMotionDemo.routeName,
+      buildRoute: (BuildContext context) => new FabMotionDemo(),
+    ),
```
might be better to just have this demo replace the existing FAB demo?

DaveShuckerow replied:
The two show off different sets of behavior.  

If we want to consolidate demos, then I think it's better to turn this into a Bottom App Bar demo, that way we can show how the two interact during the motion.  That is a change I'd prefer to keep separated from just landing these Floating Action Button changes.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,289 @@ enum _ScaffoldSlot {
+
+  /// End-aligned [FloatingActionButton], floating at the bottom of the screen.
+  /// 
+  /// This is the default alignment of [FloatingActionButton]s in Material apps.
```
apps -> applications

DaveShuckerow replied:
Done.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,289 @@ enum _ScaffoldSlot {
+///  * [FloatingActionButton], which is a circular button typically shown in the
+///    bottom right corner of the app.
+///  * [FloatingActionButtonPositioner], which the [Scaffold] uses to place the 
+///    [Scaffold.floatingActionButton] within the [Scaffold]'s layout. 
```
nice docs, btw.

DaveShuckerow replied:
:blush:

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,289 @@ enum _ScaffoldSlot {
+    return new AnimationMax<double>(
+      new ReverseAnimation(new CurveTween(curve: curve.flipped).animate(parent)),
+      new CurveTween(curve: curve).animate(parent),
+    );
```
maybe add a comment here explaining what this is trying to do? it seems like crazy magic code :-)

DaveShuckerow replied:
Done.

Hixie replied:
Using AnimationMax here won't in general work for what you describe in the comment. Suppose we change the curve to a bouncing curve that goes negative. Using AnimationMax will clamp it to zero when it should be negative.

I guess you could use a kind of Animation that adds two child animations, but what you really want is a kind of Animation that defers to one animation for a range of numbers and a second animation for a second range of numbers. That would replace the Interval in your logic here.

DaveShuckerow replied:
Done. Please take a look at the _AnimationSwap class I've introduced.

DaveShuckerow replied:
Done.  Please take a look at the _AnimationSwap.

I am unsure about making it a part of the public API because I think it's too similar to TrainHoppingAnimation.  That animation doesn't work here, I believe because I can't specify whether to maximize or minimize the animations.

I think long-term it might be better to change the train hopping animation to support this use case.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1322,16 +1727,22 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
-              geometryNotifier: _geometryNotifier,
-            ),
-          ),
+          child: new AnimatedBuilder(animation: _floatingActionButtonMoveController, builder: (BuildContext context, _) {
```
usually we prefer to type and name all the arguments, even those we don't use, so that people reading the code can tell at a glance what they are (rather than having to go look it up in the docs)

DaveShuckerow replied:
Done

Hixie commented on packages/flutter/test/material/scaffold_test.dart:
```diff
@@ -771,6 +774,96 @@ void main() {
+    });
+
+  });
+
```
i'd like to see more tests, e.g. tests of the logic that handles interrupted animations. Something else that it looks like you don't test is the precise rotation at various times in the animation cycle.

DaveShuckerow replied:
To test the progress of interrupted animations' position and scale, I can check the publicly available scaffoldgeometry listener.

To test the rotation, I need to find a way to listen to the animations themselves, which are private to the fab's container.

Does Flutter allow for fields to be marked @VisibleForTesting?  I'd like to avoid creating a new API surface just to test the animations.

DaveShuckerow replied:
I wrote a test that checks for jumps in the transition animation.  I'm not sure if there's a possibility it could flake in a test environment that runs the tests at a lower 'framerate'.  Since these are VM tests and not instrumented tests, I don't think so, but I'd like you to take a close look.

Hixie commented:
looks pretty good, just some minor issues.

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into fab-motion

Commit:
Apps -> Applications in docs

Commit:
Merge branch 'fab-motion' of github.com:DaveShuckerow/flutter into fab-motion

Commit:
Explain the scale animation.

Commit:
Name the child parameter in the animated builder

Commit:
RTL before LTR

Commit:
Wrap the AppBar in the example code

Commit:
const the fab motion demo name

Commit:
Start a test against animation jumps

Commit:
Test for jumps in the fab motion animation

Commit:
Dont initialize values to null

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// The minimum horizontal padding the [FloatingActionButton] should observe.
+  final double horizontalFloatingActionButtonPadding;
+
+  /// The [Size] of the whole [Scaffold].
```
more detail on this comment would be good. For example, does it include the size of the insets that are avoided when you use resizeToAvoidBottomPadding?

Hixie commented:
Approach looks good. I'll defer to someone else on the team to do the final detailed review.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+
+/// The geometry of the [Scaffold] before it finishes laying out.
+/// 
+/// The Scaffold passes this prelayout geometry to its [FloatingActionButtonPositioner].
```
You need to explain what "before it finishes laying out" actually means. I think what you means is that everything in [ScaffoldGeometry] has a size and a position by the FAB (of course) does not?

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+class ScaffoldPrelayoutGeometry {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const ScaffoldPrelayoutGeometry({this.bottomSheetSize, this.contentBottom, this.contentTop, this.floatingActionButtonSize, this.horizontalFloatingActionButtonPadding, this.scaffoldSize, this.snackBarSize, this.textDirection});
```
Typically we'd format a long named parameter list like this:

const Foo({
  this.a,
  this.b,
  this.c,
});

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// this should be [Size.zero].
+  final Size bottomSheetSize;
+
+  /// The height from the [Scaffold]'s top where its body ends.
```
I think this means: the vertical distance from the Scaffold's origin to the bottom of its [Scaffold.body]. 

It would be helpful to add a sentence or two about why this value is useful and why it's not bottomSheetSize.height.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// is currently showing a [BottomSheet].
+  /// 
+  /// If the [Scaffold] is not currently showing a [BottomSheet],
+  /// this should be [Size.zero].
```
In that case this value _will be_ zero, right?

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// The [Scaffold] will determine the [floatingActionButtonSize] if 
+  /// [Scaffold.floatingActionButton] is not null.
+  /// 
+  /// If [Scaffold.floatingActionButton] is null, this should be [Size.zero].
```
In that case this value _will be_ zero, right?

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// const constructors so that they can be used in const expressions.
+  const ScaffoldPrelayoutGeometry({this.bottomSheetSize, this.contentBottom, this.contentTop, this.floatingActionButtonSize, this.horizontalFloatingActionButtonPadding, this.scaffoldSize, this.snackBarSize, this.textDirection});
+
+  /// The [Size] of the [Scaffold]'s [FloatingActionButton] (if available).
```
You could boil most of the verbiage into this summary statement. Something like:  The [Size] of the [Scaffold]'s [FloatingActionButton] if [Scaffold.floatingActionButton] is not null, otherwise [Size.zero].

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// If [Scaffold.floatingActionButton] is null, this should be [Size.zero].
+  final Size floatingActionButtonSize;
+
+  /// The [Size] of the [Scaffold]'s [BottomSheet] (if available).
```
Same comment about verbiage boiling as for floatingActionButtonSize.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// The height from the [Scaffold]'s top where its body starts.
+  final double contentTop;
+
+  /// The minimum horizontal padding the [FloatingActionButton] should observe.
```
Not sure what this means, particularly the "observe" part.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// The height from the [Scaffold]'s top where its body ends.
+  final double contentBottom;
+
+  /// The height from the [Scaffold]'s top where its body starts.
```
The vertical distance from the Scaffold's origin to the top of its [Scaffold.body].

As before, it would be useful to explain why this distance is useful.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// The [Size] of the whole [Scaffold].
+  final Size scaffoldSize;
+
+  /// The [Size] of the [Scaffold]'s [SnackBar] (if available).
```
Same comment as for the earlier size fields.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  @override
+  Offset getOffset(ScaffoldPrelayoutGeometry scaffoldGeometry) {
+    // Compute the x-axis offset.
+    final double fabX = (scaffoldGeometry.scaffoldSize.width - scaffoldGeometry.floatingActionButtonSize.width) / 2;
```
We always using floating point constants in double expressions, so 2.0.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+/// To get a [ValueNotifier] for the scaffold geometry call
+/// [Scaffold.geometryOf].
+/// 
+/// For information about the [Scaffold]'s geometry that is used as part of
```
The two layout geometry classes need better explanations.  Why do both exist? When can the geometry information be used?

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -68,14 +354,14 @@ class ScaffoldGeometry {
 
-  /// The distance from the scaffold's top edge to the top edge of the
+  /// The distance from the [Scaffold]'s top edge to the top edge of the
   /// rectangle in which the [Scaffold.bottomNavigationBar] bar is being laid
```
"is being laid out"? I thought the point of this class was that everything had been laid out?

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -68,14 +354,14 @@ class ScaffoldGeometry {
   /// rectangle in which the [Scaffold.bottomNavigationBar] bar is being laid
   /// out.
   ///
   /// When there is no [Scaffold.bottomNavigationBar] set, this will be null.
```
Null if [Scaffold.bottomNavigationBar] is null.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -68,14 +354,14 @@ class ScaffoldGeometry {
   final double bottomNavigationBarTop;
 
-  /// The rectangle in which the scaffold is laying out
+  /// The rectangle in which the [Scaffold] is laying out
```
The [Scaffold.floatingActionButton]'s bounding rectangle.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -289,23 +585,24 @@ class _ScaffoldLayout extends MultiChildLayoutDelegate {
+      );
+      final Offset currentFabOffset = currentFloatingActionButtonPositioner.getOffset(currentGeometry);
+      final Offset previousFabOffset = previousFloatingActionButtonPositioner.getOffset(currentGeometry);
+      final Offset fabOffset = floatingActionButtonMotionAnimator.getOffset(begin: previousFabOffset, end: currentFabOffset, progress: floatingActionButtonMoveAnimationProgress);
```
Line breaks for the named parameters would be helpful here. 

```
 final Offset fabOffset = floatingActionButtonMotionAnimator.getOffset(
   begin: previousFabOffset, 
   end: ...
 );
```

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -442,34 +794,28 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
-      ));
-    }
-    if (_currentAnimation.status != AnimationStatus.dismissed) {
+     final List<Widget> children = <Widget>[];
```
There's an extra leading space here.

HansMuller reviewed (commented):
Mostly small stuff, overall this looks good.

HansMuller commented:
Looks like you still need to address two of @Hixie's comments
https://github.com/flutter/flutter/pull/14368/files#r173874416
https://github.com/flutter/flutter/pull/14368/files#r173874797

Commit:
Use constants, fix spacing from some of Hans' comments

Commit:
Clarify the relationship between fab positioners and prelayout geometries

Commit:
Explain the fab animmator a bit better

Commit:
Explain the animation progress in the fab animation

Commit:
Explain the animation restart better

Commit:
Explain the animation restart better

Commit:
Explain the prelayout geometry better

Commit:
Explain that height is a vertical distance.

Commit:
Explain the horizontal fab padding

Commit:
Update the scaffold size description to explain what happens when a wild keyboard appears

Commit:
Remove print statements

Commit:
Update the scaffold geometry with information about it being available at paint time.

Commit:
In one step of a transition

Commit:
Explain how the top-start fab positioner works

Commit:
Explain how the top-start fab positioner works

Commit:
Refactor the scaffold layout to just pass a padding instead of a bottom, top, start and end.

Commit:
Refactor the scaffold layout to just pass a padding instead of a bottom, top, start and end.

Commit:
Action buttons with with custom positioners.

Commit:
Add a rotation animation example.

Commit:
Use a swap animation to show swapping between two different animations.

Commit:
Use a swap animation to show swapping between two different animations.

Commit:
Add an example for the size animations.

DaveShuckerow commented on examples/flutter_gallery/lib/gallery/item.dart:
```diff
@@ -158,6 +158,13 @@ List<GalleryItem> _buildGalleryItems() {
     ),
+    new GalleryItem(
+      title: 'Floating action button motion',
+      subtitle: 'Action buttons with customized motion',
```
Yes.  I've updated this.

DaveShuckerow replied:
Yes, done.

DaveShuckerow commented on examples/flutter_gallery/lib/demo/material/fab_motion_demo.dart:
```diff
@@ -0,0 +1,111 @@
+  }
+}
+
+class _TopStartFloatingActionButtonPositioner extends FloatingActionButtonPositioner {
```
Done.

DaveShuckerow commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// Animates the scale of the [FloatingActionButton].
+  /// 
+  /// The animation should both start and end with a value of 1.0.
+  Animation<double> getScaleAnimation({@required Animation<double> parent});
```
Done.

DaveShuckerow commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// The animation should both start and end with a value of 0.0 or 1.0.
+  /// 
+  /// The animation values are a fraction of a full circle, with 0.0 and 1.0
+  /// corresponding to 0 and 360 degrees, while 0.5 corresponds to 180 degrees.
```
Done.

DaveShuckerow reviewed (commented):
Responding to everything, including Ian's animation comment.

DaveShuckerow commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1304,34 +1832,32 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
-    assert(endPadding != null);
-
+    // The minimum insets for contents of the Scaffold to keep visible.
+    final EdgeInsets minInsets = mediaQuery.padding.copyWith(
```
@Hixie @HansMuller 

Please take a look at this change.  We discussed it separately, but brief recap:


> What's happening is that the top, end and bottom of mediaQuery.padding are all getting passed to the scaffold layout
But with start-aligned fabs, we really need the start padding as well
and so at this point it's easier to just pass the padding directly and find the start and end padding in the relevant positioner

HansMuller commented on examples/flutter_gallery/lib/demo/material/fab_motion_demo.dart:
```diff
@@ -0,0 +1,152 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
Should be 2018

DaveShuckerow replied:
Done

HansMuller commented on examples/flutter_gallery/lib/demo/material/fab_motion_demo.dart:
```diff
@@ -0,0 +1,152 @@
+  }
+}
+
+
```
There's an extra blank line here

DaveShuckerow replied:
Done

HansMuller commented on examples/flutter_gallery/lib/demo/material/fab_motion_demo.dart:
```diff
@@ -0,0 +1,152 @@
+        ),
+      ),
+    );
+    return scaffold;
```
Just `return new Scaffold` above.

DaveShuckerow replied:
Done.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// Centered [FloatingActionButton], floating at the bottom of the screen.
+  static const FloatingActionButtonPositioner centerFloat = const _CenterFloatFabPositioner();
+
+  /// Places the [FloatingActionButton] based on the [Scaffold]'s layout.
```
NICE

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,410 @@ enum _ScaffoldSlot {
+  ///   @override
+  ///   Animation<double> getScaleAnimation({@required Animation<double> parent}) {
+  ///     // The animations will cross at value 0, and the train will return to 1.0.
+  ///     return new TrainHoppingAnimation(
```
Nice example. Now I (finally) understand what a train hopping animation is :-).

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,410 @@ enum _ScaffoldSlot {
+///    another.
+///  * [ScaffoldPrelayoutGeometry], the geometry that 
+///    [FloatingActionButtonPositioner]s use to position the [FloatingActionButton].
+abstract class FloatingActionButtonPositioner {
```
scaffold.dart is pretty big, could this class (et al) live in fab_animation.dart?

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,410 @@ enum _ScaffoldSlot {
+  @override
+  T get value => parent.value < swapThreshold ? first.value : next.value;
+}
+
```
Extra blank line

DaveShuckerow replied:
Done

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,410 @@ enum _ScaffoldSlot {
+}
+
+
+/// The geometry of the [Scaffold] after it all of its contents except for the
```
all its contents have been laid out, execpt ...

DaveShuckerow replied:
Done

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,410 @@ enum _ScaffoldSlot {
+/// 
+/// The [Scaffold] passes this prelayout geometry to its
+/// [FloatingActionButtonPositioner], which produces an [Offset] that the 
+/// [Scaffold] uses to finish laying out the [FloatingActionButton].
```
micro-nit: Scaffold just uses the offset to position the FAB, right?

DaveShuckerow replied:
Yes.  Clarified.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,410 @@ enum _ScaffoldSlot {
+  /// 
+  /// If the [Size] of the [Scaffold]'s contents is modified by values such as 
+  /// [Scaffold.resizeToAvoidBottomPadding] or the keyboard opening, then the
+  /// [scaffoldSize] here will not reflect those changes.
```
You can leave out the word "here" here.

DaveShuckerow replied:
Done here.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -55,10 +60,410 @@ enum _ScaffoldSlot {
+/// 
+/// The ScaffoldGeometry is only available during the paint phase, because
+/// its value is computed during the animation and layout phases prior to painting.
+/// 
```
It would be helpful to briefly explain what a typical use case for ScaffoldGeometry is.

DaveShuckerow replied:
Done.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -193,19 +596,25 @@ class _ScaffoldGeometryNotifier extends ChangeNotifier implements ValueListenabl
+    @required this.currentFloatingActionButtonPositioner,
+    @required this.floatingActionButtonMoveAnimationProgress,
+    @required this.floatingActionButtonMotionAnimator,
+  }) : assert(previousFloatingActionButtonPositioner != null), assert(currentFloatingActionButtonPositioner != null);
```
We usually put asserts like this on separate lines

DaveShuckerow replied:
Done.

HansMuller reviewed (approved):
Just a few very minor cleanups and maybe a little file management. This looks good!

LGTM

Commit:
2018 copyright

Commit:
Extra empty line

Commit:
Return new Scaffold

Commit:
Merge branch 'fab-motion' of github.com:DaveShuckerow/flutter into fab-motion

Commit:
Extra blank line fix

Commit:
All its contents have been laid out

Commit:
Position the fab

Commit:
Explain what the scaffold geometry is for.

Commit:
Move asserts to different lines

Commit:
The scaffoldsize will not

Commit:
Initial rename of FabPositioners to FloatingActionButtonLocation

Commit:
Rename comments in example to refer to location instead of positioner.

Commit:
Rename fabpositioner to location in tests and in the scaffold field

Commit:
Finish removing references to positioner in scaffold code.

Commit:
Split the fab location and animation out into a separate file.

Commit:
Make things more private

Commit:
Import foundation instead of meta

Commit:
Const curve instead of final.

HansMuller commented:
Thanks for breaking out floating_action_button_location.dart; made scaffold.dart a little easier to read.

Still LGTM!

DaveShuckerow commented:
Once the tests pass, I will merge this commit.  Thanks for the review, Hans and Ian!

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -430,7 +593,43 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
+      parent: _previousController,
+      curve: Curves.easeIn,
+    );
+    final Animation<double> previousExitRotationAnimation = new Tween<double>(begin: 1.0, end: 1.0).animate(
```
am I missing something or is this animation always going to say 1.0?

Hixie replied:
https://github.com/flutter/flutter/pull/21535
